software/controller/test_libs/mocks/mock_spi_chain.h

```diff
@@ -0,0 +1,155 @@
+/* Copyright 2022, RespiraWorks
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#pragma once
+
+#include "spi_chain.h"
+
+namespace SPI {
+
```

Comment:
I am not sure I like how this is set up. So we have the `DaisyChain` class, which has some HAL-specific functionality, but also some generic functionality. And now we derive from it mainly so we can shadow the functions that are too HAL-coupled and manipulate the parent class partially through the protected members? This is a bit hard to follow and it is not very clear which "cuts" of the code are destined to be tested, and which - not.

I would prefer that there were an `spi_chain.h` that only defined an abstract interface without assumptions about implementation, and then we have 2 derived classes that implement the pure virtual as either HAL or mock. Or it could be in a different layer and the class can be given a back-end as template type similar to how it was done with sensors.

Reply:
I'll see what I can do about it. The root of the problem is that I am deriving `SPI::DaisyChain` from `SPI::Channel`. The highly hal-coupled thing in there is `SPI::Channel`. The `SPI::MockChain` is indeed shadowing from its parent class, but this is stuff it inherited from `SPI::Channel`, which means making `SPI::DaisyChain` abstract does not solve the problem.
A solution I can see is to make `SPI::DaisyChain` have a member pointing to its`SPI::Channel` and have that be abstract, instead of inheriting from it.

Reply:
Yes, making Channel abstract sounds like it may be the right strategy for this.